var documenterSearchIndex = {"docs":
[{"location":"#fastPHASE.jl","page":"Home","title":"fastPHASE.jl","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"fastPHASE.jl is a Julia wrapper of the popular fastPHASE genetics program, designed for haplotype reconstruction and estimating missing genotypes from population data.","category":"page"},{"location":"","page":"Home","title":"Home","text":"The methodology is described in the following paper","category":"page"},{"location":"","page":"Home","title":"Home","text":"Scheet, P and Stephens, M (2006). A fast and flexible statistical model for large-scale population genotype data: applications to inferring missing genotypes and haplotypic phase. Am J Hum Genet. https://doi.org/10.1086/502802","category":"page"},{"location":"man/api/#API","page":"API","title":"API","text":"","category":"section"},{"location":"man/api/","page":"API","title":"API","text":"Here is a list of available function calls. A detailed description can be found below. ","category":"page"},{"location":"man/api/#Index","page":"API","title":"Index","text":"","category":"section"},{"location":"man/api/","page":"API","title":"API","text":"Pages = [\"api.md\"]","category":"page"},{"location":"man/api/#Functions","page":"API","title":"Functions","text":"","category":"section"},{"location":"man/api/","page":"API","title":"API","text":"  fastphase_estim_param\n  process_fastphase_output","category":"page"},{"location":"man/api/#fastPHASE.fastphase_estim_param","page":"API","title":"fastPHASE.fastphase_estim_param","text":"fastphase_estim_param(xdata; ...)\n\nRuns fastPHASE to estimate r, θ, α parameters. Will run different initial EM starts in parallel if Julia is started with multiple threads. θ (emission) probabilities) values will automatically be flipped so that θ[i, j] is the probability of observing allele A2 (usually major) in the PLINK bam file at SNP i of haplotype j.\n\nInputs\n\nxdata: A String for binary PLINK file (without .bed/bim/fam extensions)      or a SnpData (see SnpArrays.jl). \n\nOptional inputs\n\nn: Number of samples used to fit HMM in fastPHASE. Defaults to sample size in xdata. T: Number of different initial conditions for EM. Different initial conditions     will be run in parallel in Threads.nthreads() number of threads. K: Number of haplotype clusters. Defaults to 12 C: Number of EM iterations before convergence. Defaults to 10. outfile: Extension of output alpha, theta, and r file names. Defaults to fastphase_out outdir: Output directory. By default all output will be stored in new folder in      knockoffs in the current directory. fastphase_infile: Filename of fastPHASE's input, which is the decompressed PLINK     genotypes readable by fastPHASE. Defaults to fastphase.inp. \n\n\n\n\n\n","category":"function"},{"location":"man/api/#fastPHASE.process_fastphase_output","page":"API","title":"fastPHASE.process_fastphase_output","text":"process_fastphase_output(filename; [T])\n\nReads r, θ, α into memory, averaging over T simulations. θ (emission probabilities) must be flipped sometimes depending on which allele was defined as \"allele 1\". fastPHASE simply uses whichever allele was observed first in sample 1 haplotype 1. Thus, in sample 1, genotypes that start with \"10\" or \"1?\" must be flipped, and this info is provided in the \"_origchars\" file. \n\nInputs\n\nfilename: Path to the rhat.txt, thetahat.txt, alphahat.txt and     _origchars files are stored. E.g. Use out     if your output files are out_rhat.txt, out_thetahat.txt, out_alphahat.txt T: Number of different runs excuted by fastPHASE. This is the number of different      initial conditions used for EM algorithm. All files rhat.txt, thetahat.txt,     alphahat.txt would therefore have T × p rows\n\n\n\n\n\n","category":"function"}]
}
